code/contracts/ONEWallet.sol

```diff
@@ -64,30 +65,33 @@ contract ONEWallet {
         (uint32 ct, bool completed) = _findCommit(hash);
         require(ct == 0 && !completed, "Commit already exists");
         Commit memory nc = Commit(hash, uint32(block.timestamp), false);
-        commits.push(nc);
+        commits.push(nc); // IH: would not it be better to have mapping of commits?
     }
 
     // TODO (@polymorpher): during reveal, reject if index corresponds to a timestamp that is same as current block.timestamp, so we don't let the client accidentally leak EOTP that is still valid at the current time, which an attacker could use to commit and reveal a new transaction. This introduces a limitation that there would be a OTP_INTERVAL (=30 seconds by default) delay between commit and reveal, which translates to an unpleasant user experience. To fix this, we may slice OTP_INTERVAL further and use multiple EOTP per OTP_INTERVAL, and index the operations within an interval by an operation id. For a slice of size 16, we will need to generate a Merkle Tree 16 times as large. In this case, we would reject transactions only if both (1) the index corresponds to a timestamp that is same as current block.timestamp, and, (2) the operation id is not less than the id corresponding to the current timestamp's slice
+    // IH: what do you state above just supports HOTPs instead of TOTPs, besides other advantages.
+    // IH: combining index with nonce is unnecessary complicated to me. I would split them. 
     function revealTransfer(bytes32[] calldata neighbors, uint32 indexWithNonce, bytes32 eotp, address payable dest, uint256 amount) external
     isCorrectProof(neighbors, indexWithNonce, eotp)
     returns (bool)
     {
-        bytes memory packedNeighbors = _pack(neighbors);
+        bytes memory packedNeighbors = _pack(neighbors); // IH: no need to put Merkle proof into the hash (it is either correct and matches to OTP or not, hence integrity is implicit)
```

Comment:
That's true. However, if we remove the `neighbors` from the packed bytes computation, the only issue seems to be that an attacker could commit the exact same transaction ahead of the user, resulting the user receiving an annoying rejection message "commit already exists" when the user truly commits it. This may result in a bad user experience. Since only neighbors[0] is secret, we could keep that inside the `packed` instead of the whole `neighbors`. This way the attacker would have no way of committing the same transaction ahead of the user since they would not know the value of `neighbors[0]`

Reply:
You are right. I somehow thought EOTP would be the same for two different nonces. They are not. So EOTP alone is sufficient. But based on our discussion (outside this PR) we are going to rewrite the logic for commit-reveal anyway to cure one of the attacks.